### PR TITLE
Add selective data export to CLI

### PR DESCRIPTION
**Pull request by Pytal**
Close https://github.com/nextcloud/user_migration/issues/94

```
Description:
  Export a user.

Usage:
  user:export [options] [--] [<user> [<folder>]]

Arguments:
  command               The command to execute
  user                  user to export
  folder                local folder to export into

Options:
  -l, --list[=LIST]     List the available data types, pass --list=full to show more information [default: false]
  -t, --types=TYPES     Comma-separated list of data type ids, pass --types=none to only export base user data [default: false]
  -h, --help            Display this help message
```

### CODE REVIEW

File: lib/Command/Export.php [hunk 1]
```diff
@@ -54,43 +61,127 @@ protected function configure(): void {
...
+					$selectedMigrators = $types;
+				}
+			}
+		}
 
-		if (!$userObject instanceof IUser) {
-			$output->writeln("<error>Unknown user " . $input->getArgument('user') . "</error>");
+		$uid = $input->getArgument('user');
+		if (empty($uid)) {
+			$io->warning('Missing user argument');
```
Comment (come-nc) [inline lib/Command/Export.php:153]: Should be error?

File: lib/Command/Export.php [hunk 2]
```diff
@@ -54,43 +61,127 @@ protected function configure(): void {
...
+		$types = $input->getOption('types');
+		if ($types !== false) {
+			$types = explode(',', $types);
+			if ($types !== false) {
+				if (count($types) === 1 && reset($types) === 'none') {
+					$selectedMigrators = [];
+				} else {
+					foreach ($types as $id) {
+						if (!in_array($id, array_map(fn (IMigrator $migrator) => $migrator->getId(), $migrators), true)) {
+							$io->warning("Invalid type: \"$id\"");
```
Comment (come-nc) [inline lib/Command/Export.php:142]:
```suggestion
							$io->error("Invalid type: \"$id\"");
```

File: lib/Command/Export.php [hunk 3]
```diff
@@ -54,43 +61,127 @@ protected function configure(): void {
...
+
+		$user = $this->userManager->get($uid);
+		if (!$user instanceof IUser) {
+			$io->error("Unknown user <$uid>");
 			return 1;
 		}
 
+		$folder = $input->getArgument('folder');
+		if (empty($folder)) {
+			$io->warning('Missing folder argument');
```
Comment (come-nc) [inline lib/Command/Export.php:165]:
```suggestion
			$io->error('Missing folder argument');
```

**Review by come-nc — COMMENTED** [review]
Did not test, will do tomorrow hopefully.

**Comment by come-nc** [comment]
The help does not match behavior:
```
Usage:
  user:export [options] [--] [<user> [<folder>]]
```
```
occ user:export bob
                                                                                                                        
 [ERROR] Missing folder argument       
```

This cannot be fixed easily I think

**Comment by come-nc** [comment]
(tested and works, you can merge)

**Comment by Pytal** [comment]
> The help does not match behavior:
> …
> This cannot be fixed easily I think

Not with the existing symfony console framework, errors are thrown as needed to inform the user

> How hard would it be to add an `--ouput` option like core commands to get the list in json format?

Will be tracked in https://github.com/nextcloud/user_migration/issues/170

**Pull request merged**